CONTRIBUTING.md

```diff
@@ -1,12 +1,100 @@
+You must first create your own fork of the project and submit your changes to a branch.
+You can then raise a pull request targeting `oneapi-src/unified-memory-framework:main`.
+Please try to keep commits neat and in order - please squash your commits to
+include only relevant ones (no "fixes after review" or similar).
+
+Before your pull request is merged it *must* pass all jobs in the GitHub
+Actions workflows and *must* be reviewed by no less than two reviewers.
+
+### Building and testing
+
+Building commands can be found in the top-level Readme file - section
+["Build"](./README.md#build).
+
+Before committing you should test locally if all tests and checks pass.
+When project is built, enter the build directory and execute:
+
+```bash
+$ ctest --output-on-failure
+```
+
+Any test's failure will produce error log.
+
+To enable additional checks (including `-Werror` / `/WX` compilation flag), switch on CMake flag
+`UMF_DEVELOPER_MODE`. To read more about all available CMake options please see
+["CMake standard options"](./README.md#cmake-standard-options) section in the top-level Readme.
+
+### Code style
+
+We use `clang-format` to verify and apply code style changes. To see all rules we require,
+please take a look at `.clang-format` file in the root directory of this repository.
+
+To enable code style checks and re-formatting, CMake option `UMF_FORMAT_CODE_STYLE` has to
+be switched on. You'll then have two additional commands available.
+
+To verify correct coding style of your changes execute (assuming `build` is your build directory):
+
+```bash
+$ cmake -B build -DUMF_FORMAT_CODE_STYLE=ON
+$ cmake --build build --target clang-format-check
+```
+
+If any issues were found, apply proper formatting via command:
+
+```bash
+$ cmake --build build --target clang-format-apply
+
+# Remember to review introduced changes
+```
+
+**NOTE**: We use specific clang-format version - **15.0** is required. It can be installed,
+e.g., with command: `python -m pip install clang-format==15.0.7`.
+
+<!--
+### Naming convention
+TODO: add this section and re-format whole codebase to use such convention
+-->
 
 ### License
 
-<PROJECT NAME> is licensed under the terms in [LICENSE]<link to license file in repo>. By contributing to the project, you agree to the license and copyright terms therein and release your contribution under these terms.
+Unified Memory Framework is licensed under the terms in [LICENSE](./LICENSE.TXT). By contributing to the project,
+you agree to the license and copyright terms therein and release your contribution under these terms.
 
 ### Sign your work
 
-Please use the sign-off line at the end of the patch. Your signature certifies that you wrote the patch or otherwise have the right to pass it on as an open-source patch. The rules are pretty simple: if you can certify
+Please use the sign-off line at the end of the patch. Your signature certifies that you wrote the patch
```

Comment:
I thought that signing-off wasn't meant to be a requirement in this project, @bratpiorka ?

Reply:
It's still not a requirement - it's merely an ask 😉 Also, for people in our organization this may not be needed, but perhaps for outside contributors, it's a good thing to ask...?

Reply:
I would remove it now. I am not sure if even my approvals are signed. This is also not tested in CI. IMO this whole file contains not questions, but requirements that would fail CI if not met.

Reply:
Again - could we test this? Even if not now, then in the future? If we leave it, the whole UMF team should respect it and we should not merge any non-signed-off code. But if it is a 'nice to have', I would not mention it to avoid unnecessary misunderstandings and discussions.

Reply:
heh, we had a clause like this in all(?) our previous repos, and I cannot remember a single discussion about it 😉 
We can test anything, including this (but I hope there's a tool for that 😉 ).

// examples: [PMDK](https://github.com/pmem/pmdk/blob/master/CONTRIBUTING.md) / [pmemkv](https://github.com/pmem/pmemkv/blob/master/CONTRIBUTING.md)

That said, we can use similar wording to one in PMDK, to be more flexible about it:

>In case of any doubt, the gatekeeper may ask you to certify the above in writing, i.e. via email or by including a Signed-off-by: line at the bottom of your commit comments.

Reply:
"Anecdotal evidence" :D I'm ok with "In case of any doubt" form. If we want to make it stronger we have to follow it + add tests.

Reply:
Rewritten.